code/controllers/master_controller.dm

```diff
@@ -23,19 +23,20 @@ var/global/datum/controller/game_controller/master_controller = new()
 			init_subtypes(/datum/subsystem, subsystems)
 
 		master_controller = src
+	calculateGCD()
 
+/datum/controller/game_controller/proc/calculateGCD()
 /*
 calculate the longest number of ticks the MC can wait between each cycle without causing subsystems to not fire on schedule
-Note: you can set the datum's defined processing_interval to some integer to set an -absolute- minimum wait duration.
 */
 	var/GCD
 	for(var/datum/subsystem/SS in subsystems)
 		if(SS.wait)
 			GCD = Gcd(SS.wait, GCD)
 	GCD = round(GCD)
-	if(GCD > processing_interval)
-		processing_interval = GCD
-
+	if(GCD < 1)
+		GCD = 1
+	processing_interval = GCD

```

Comment:
This was from carnmc mostly, i just made into a proc instead of being inside mc's new.

What this does, is sets the mc's sleep rate to be the lowest it needs to be to hit every task on time. Because of the fact that subsystems can change how often they run, this is dynamically updated only when it matters.

If every subsystem is either 5ds, 10ds or 20ds, than the mc only needs to run every 5ds. but if one of those is 7ds, than the mc needs to run every 1ds to be able to hit that 7ds one on time.

Its set to 1 because a note in carnmc on subsystems.dm says wait should only be positive whole numbers. I'm not quite at the point where I understand carnmc well enough to solve issues with partial numbers, and the changes I've made to carnmc are already borderlining the scope of what should be in one pr.

Anywho, the only thing new I've coded in this pr is hot spot pooling and the logic around changing the wait of a subsystem, everything else was just moving existing code around and generalizing existing code to work with dynamic fastmos.